src/webserver/webserver.c

```diff
@@ -866,8 +868,56 @@ void http_terminate(void)
 	// Free admin_api_uri path
 	if(admin_api_uri != NULL)
 		free(admin_api_uri);
-	
+
 	// Free login_uri path
 	if(login_uri != NULL)
 		free(login_uri);
 }
+
+static void restart_http(void)
+{
+	// Stop the server
+	http_terminate();
+
+	// Reinitialize the webserver
+	http_init();
+}
+
+void *webserver_thread(void *val)
+{
+	(void)val;
+	// Set thread name
+	prctl(PR_SET_NAME, thread_names[WEBSERVER], 0, 0, 0);
+
+	// Initial delay until we check the certificate for the first time
+	thread_sleepms(WEBSERVER, 2000);
+
+	while(!killed)
+	{
+		// Check if the certificate is about to expire soon
+		const enum cert_check status = cert_currently_valid(config.webserver.tls.cert.v.s, 2);
+
+		if(status == CERT_EXPIRES_SOON &&
+		   config.webserver.tls.validity.v.ui > 0 &&
+		   is_pihole_certificate(config.webserver.tls.cert.v.s))
+		{
+			log_info("TLS certificate at %s is about to expire soon, generating new one",
+			         config.webserver.tls.cert.v.s);
+			generate_certificate(config.webserver.tls.cert.v.s, false,
+			                     config.webserver.domain.v.s,
+			                     config.webserver.tls.validity.v.ui);
+
+			log_info("Restarting HTTP server");
+			restart_http();
+
+			log_info("Done. The new certificate is valid for %u days",
+			         config.webserver.tls.validity.v.ui);
+		}
```

Comment:
Should we add an `else` here in case `is_pihole_certificate=false` to log something like.

log_info("TLS certificate at %s is going to expire soon but not issued by Pi-hole. Leaving untouched",
			         config.webserver.tls.cert.v.s);